File: backend/app/domains/wealth/schemas/portfolio.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class PortfolioSummary(BaseModel):
    profile: str
    snapshot_date: date | None = None
    cvar_current: Decimal | None = None
    cvar_limit: Decimal | None = None
    cvar_utilized_pct: Decimal | None = None
    trigger_status: str | None = None
    regime: str | None = None
    core_weight: Decimal | None = None
    satellite_weight: Decimal | None = None
    computed_at: datetime | None = None


class PortfolioSnapshotRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    snapshot_id: uuid.UUID
    profile: str
    snapshot_date: date
    weights: dict[str, Any]
    fund_selection: dict[str, Any] | None = None
    cvar_current: Decimal | None = None
    cvar_limit: Decimal | None = None
    cvar_utilized_pct: Decimal | None = None
    trigger_status: str | None = None
    consecutive_breach_days: int = 0
    regime: str | None = None
    core_weight: Decimal | None = None
    satellite_weight: Decimal | None = None
    regime_probs: dict[str, Any] | None = None
    # Bayesian credible interval bounds (negative = loss, same sign as cvar_current).
    # cvar_lower_5 is the 5th posterior percentile (worst-case tail);
    # cvar_upper_95 is the 95th percentile (best-case tail within the interval).
    # Both are None until the Bayesian CVaR worker has run for the snapshot date.
    cvar_lower_5: Decimal | None = None
    cvar_upper_95: Decimal | None = None
    computed_at: datetime | None = None


class RebalanceRequest(BaseModel):
    trigger_reason: str | None = Field(None, max_length=5000)


class RebalanceEventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="ignore")

    event_id: uuid.UUID
    profile: str
    event_date: date
    event_type: str
    trigger_reason: str | None = None
    weights_before: dict[str, Any] | None = None
    weights_after: dict[str, Any] | None = None
    cvar_before: Decimal | None = None
    cvar_after: Decimal | None = None
    status: str
    approved_by: str | None = None
    notes: str | None = None
    created_at: datetime


class RebalanceApproveRequest(BaseModel):
    notes: str | None = Field(None, max_length=5000)


# ── Position & Performance schemas ──────────────────────────


class PositionDetail(BaseModel):
    """Single position in a model portfolio with cost basis and live pricing."""

    model_config = ConfigDict(from_attributes=True, extra="ignore")

    instrument_id: uuid.UUID
    ticker: str | None = None
    name: str
    asset_class: str = ""
    currency: str = "USD"
    weight: Decimal = Field(..., description="Allocation weight 0.0-1.0")
    block_id: str | None = None

    # Price data (from nav_timeseries)
    last_price: Decimal | None = Field(None, description="Latest NAV/price")
    previous_close: Decimal | None = Field(None, description="Previous trading day close")
    price_date: date | None = None

    # Computed fields (populated by service layer)
    position_value: Decimal | None = Field(None, description="weight * portfolio_nav")
    intraday_pnl: Decimal | None = Field(None, description="(last_price - prev_close) * weight * portfolio_nav / last_price")
    intraday_pnl_pct: Decimal | None = Field(None, description="(last_price - prev_close) / prev_close * 100")


class PerformancePoint(BaseModel):
    """Single data point in a portfolio performance time series."""

    nav_date: date
    nav: Decimal
    daily_return: Decimal | None = None
    cumulative_return: Decimal | None = None
    benchmark_nav: Decimal | None = None
    benchmark_cumulative_return: Decimal | None = None


class PortfolioPerformanceSeries(BaseModel):
    """Historical NAV and return series for charting."""

    portfolio_id: uuid.UUID
    profile: str
    inception_date: date | None = None
    inception_nav: Decimal = Decimal("1000")
    benchmark_name: str | None = None
    series: list[PerformancePoint]
    as_of: date


# ── Drift Report schemas ──────────────────────────────────────


class BlockDriftRead(BaseModel):
    """Per-block drift between current and target weights."""

    block_id: str
    current_weight: float
    target_weight: float
    absolute_drift: float
    relative_drift: float
    status: str  # "ok" | "maintenance" | "urgent"


class DriftReportRead(BaseModel):
    """Block-level allocation drift report for a portfolio profile."""

    profile: str
    as_of_date: date
    blocks: list[BlockDriftRead]
    max_drift_pct: float
    overall_status: str  # "ok" | "maintenance" | "urgent"
    rebalance_recommended: bool
    estimated_turnover: float


class LiveDriftResponse(BaseModel):
    """Response for GET /model-portfolios/{id}/drift/live.

    Computes block-level drift using live NAV prices from nav_timeseries
    instead of stale PortfolioSnapshot weights.
    """

    portfolio_id: str
    profile: str
    as_of: date
    total_aum: float
    blocks: list[BlockDriftRead]
    max_drift_pct: float
    overall_status: str  # "ok" | "maintenance" | "urgent"
    rebalance_recommended: bool
    estimated_turnover: float
    latest_nav_date: date | None = None


# ── Monitoring Alert schemas ──────────────────────────────────


class AlertRead(BaseModel):
    """A single monitoring alert."""

    alert_type: str
    severity: str
    title: str
    detail: str
    entity_id: str | None = None
    entity_type: str | None = None


class AlertBatchRead(BaseModel):
    """Batch of alerts from a monitoring scan."""

    alerts: list[AlertRead]
    scanned_at: str
    organization_id: str

File: frontends/wealth/src/lib/types/portfolio.ts
/** Portfolio (tactical) domain types — maps 1:1 to backend schemas. */

export interface PortfolioSummary {
	profile: string;
	snapshot_date: string | null;
	cvar_current: string | null;
	cvar_limit: string | null;
	cvar_utilized_pct: string | null;
	trigger_status: string | null;
	regime: string | null;
	core_weight: string | null;
	satellite_weight: string | null;
	computed_at: string | null;
}

export interface PortfolioSnapshot {
	snapshot_id: string;
	profile: string;
	snapshot_date: string;
	weights: Record<string, number>;
	fund_selection: Record<string, unknown> | null;
	cvar_current: string | null;
	cvar_limit: string | null;
	cvar_utilized_pct: string | null;
	trigger_status: string | null;
	consecutive_breach_days: number;
	regime: string | null;
	core_weight: string | null;
	satellite_weight: string | null;
	regime_probs: Record<string, number> | null;
	cvar_lower_5: string | null;
	cvar_upper_95: string | null;
	computed_at: string | null;
}

export interface StrategicAllocation {
	allocation_id: string;
	profile: string;
	block_id: string;
	target_weight: number;
	min_weight: number;
	max_weight: number;
	risk_budget: number | null;
	rationale: string | null;
	approved_by: string | null;
	effective_from: string;
	effective_to: string | null;
	created_at: string;
}

export interface TacticalPosition {
	position_id: string;
	profile: string;
	block_id: string;
	overweight: number;
	conviction_score: number | null;
	signal_source: string | null;
	rationale: string | null;
	valid_from: string;
	valid_to: string | null;
	created_at: string;
}

export interface EffectiveAllocation {
	profile: string;
	block_id: string;
	strategic_weight: number | null;
	tactical_overweight: number | null;
	effective_weight: number | null;
	min_weight: number | null;
	max_weight: number | null;
}

export interface RebalanceEvent {
	event_id: string;
	profile: string;
	event_date: string;
	event_type: string;
	trigger_reason: string | null;
	weights_before: Record<string, number> | null;
	weights_after: Record<string, number> | null;
	cvar_before: string | null;
	cvar_after: string | null;
	status: string;
	approved_by: string | null;
	notes: string | null;
	created_at: string;
}

/** Editable weight row for the allocation editor */
export interface EditableWeight {
	block_id: string;
	weight: number;
	min_weight: number;
	max_weight: number;
	strategic_weight: number;
}
